CONTRIBUTING.md

```diff
@@ -29,3 +29,125 @@ Recommendations and requirements for how to best contribute to Parquet. We striv
+encouraged.
+
+The general steps for adding features to the format are as follows:
+
+1. Discuss changes on on the developer mailing list (dev@parquet.apache.org).
+   Often times it is helpful to link to a draft pull request to make the
+   discussion concrete. This step is complete when there lazy consensus. Part
+   of the consensus is whether it sufficient to provide 2 working
+   implementations as outlined in step 2 or if demonstration of the feature
+   with a down-stream query engine is necessary to justify the feature (e.g.
+   demonstrate performance improvements in Arrow's DataSet library or
+   Apache Data Fusion).
+
+2. Once a change has lazy consensus two implementations of the feature
+   demonstrating interopability must also be provided.  One implementation MUST
+   be [parquet-java](http://github.com/apache/parquet-java).  It is preferred
+   that the second implementation be
+   [parquet-cpp](https://github.com/apache/arrow) or
+   [parquet-rs](https://github.com/apache/arrow-rs), however at the discretion
+   of the PMC any open source Parquet implementation may be acceptable.
+   Implementations whose contributors actively participate in the community
+   (e.g. keep their feature matrix up-to-date on parquet-site) are more likely
+   to be considered. If discussed as a requirement in step one demonstration
+   of integration with a query engine is also required for this step.
+
+Unless otherwise discussed, it is expected the implementations will develop from
+the main branch (i.e. backporting is not expected).
+
+3. After the first two steps are complete a formal vote is held on the Parquet
+   mailing list to officially ratify the feature.  After the vote passes the
+   format change is merged into the parquet-format repository and it is expected
+   the changes from step 2 will also be merged soon after. Before merging into
+   Parquet-java a parquet-format release must be performed.
+
+#### General guidelines/preferences on additions.
+
+1. To the greatest extent possible changes should have an option for forwards
+   compatibility (old readers can still read files).
+2. New encodings should be fully specified in this repository and ideally not
+   rely on an external dependencies for implementation (i.e. Parquet is the
+   source of truth for the encoding).
+3. New compression mechanisms must have a pure Java implementation that can be
+   used as dependency in parquet-java.
+
+### Releases
+
+The Parquet community aims to do releases of the format package only as needed
+when new features are introduced. If multiple new features are being proposed
+simultaneously some features might be consolidated into the same release.
+Guidance is provided below on when implementations should enable features added
+to the specification.  Due to confusion in the past over parquet versioning it
+is not expected that there will be a 3.0 release of the specification in the
+foreseeable future.
```

Comment:
Given there seems to be confusion about what even constitutes a parquet "release" maybe we simply should say that the parquet-format has no overall release version identifier and instead is a collection of individual features?

Or maybe we should give it a version date ("parquet-format as of 2024-06-07" etc) 🤔

Reply:
I think at least for parquet-java we need a version identifier compatible with Maven because I think that is how it is consumed.  We can maybe discuss separately just inlining the generated code there as well, as i think that is typically what other distributions do?

Reply:
CC @wgtmac thoughts on this?

Reply:
Thinking more about this, I think releases are also potentially helpful for any third-party java implementations beyond parquet java.  I don't have strong feelings here.

Reply:
Although the version identifier of parquet-format is compatible with Maven, it looks much weaker than usual Java libraries. IMO, a new version of parquet-format usually includes a set of new feature additions and format clarification/correction, meaning that a new version should be a good superset of previous versions. Therefore we should always stick to the latest version of parquet-format wherever possible. An exception is the deprecation of features to the format where a major version bump is required.